Drivers/imu/src/icm42688.cpp

```diff
@@ -0,0 +1,163 @@
+#define UB0_REG_TEMP_DATA1 0x1D
+
+#define BIT_READ 0x80
+
+//Scale Factors (refer to page 11-12 of https://product.tdk.com/system/files/dam/doc/product/sensor/mortion-inertial/imu/data_sheet/ds-000347-icm-42688-p-v1.6.pdf)
+
+#define GYRO_SENSITIVITY_2000DPS 16.4           //Currently in Primary Use
+#define GYRO_SENSITIVITY_1000DPS 32.8
+#define GYRO_SENSITIVITY_500DPS 65.5
+#define GYRO_SENSITIVITY_250DPS 131
+#define GYRO_SENSITIVITY_125DPS 262
+#define GYRO_SENSITIVITY_62_5DPS 524.3
+#define GYRO_SENSITIVITY_31_25DPS 1048.6
+#define GYRO_SENSITIVITY_15_625PS 2097.2
+
+#define ACCEL_SENSITIVITY_2G 16384 
+#define ACCEL_SENSITIVITY_4G 8192
+#define ACCEL_SENSITIVITY_8G 4096
+#define ACCEL_SENSITIVITY_16G 2048              //Currently in Primary Use
+
+ICM42688::ICM42688(SPI_HandleTypeDef * spi_handle, GPIO_TypeDef * cs_gpio_port, uint16_t cs_pin) {
+    SPI_HANDLE = spi_handle;
+    CS_GPIO_PORT = cs_gpio_port;
+    CS_PIN = cs_pin;
+}
+
+void ICM42688::readRegister(uint8_t sub_address, uint8_t count, uint8_t * dest) {
+    //Set read bit for register address
+    uint8_t tx = sub_address | BIT_READ;
+
+    //Dummy transmit and receive buffers
+    uint8_t dummy_tx[count];
+    uint8_t dummy_rx;
+
+    //Initialize values to 0
+    memset(dummy_tx, 0, count * sizeof(dummy_tx[0]));
+
+    HAL_GPIO_WritePin(CS_GPIO_PORT, CS_PIN, GPIO_PIN_RESET);
+
+    HAL_SPI_TransmitReceive(SPI_HANDLE, &tx, &dummy_rx, 1, HAL_MAX_DELAY);
+    HAL_SPI_TransmitReceive(SPI_HANDLE, dummy_tx, dest, count, HAL_MAX_DELAY);
+
+    HAL_GPIO_WritePin(CS_GPIO_PORT, CS_PIN, GPIO_PIN_SET);
+}
+
+void ICM42688::writeRegister(uint8_t sub_address, uint8_t data) {
+    //Prepare transmit buffer
+    uint8_t tx_buf[2] = {sub_address, data};
+
+    HAL_GPIO_WritePin(CS_GPIO_PORT, CS_PIN, GPIO_PIN_RESET);
+
+    HAL_SPI_Transmit(SPI_HANDLE, tx_buf, sizeof(tx_buf), HAL_MAX_DELAY);
+    
+    HAL_GPIO_WritePin(CS_GPIO_PORT, CS_PIN, GPIO_PIN_SET);
+}
+
+void ICM42688::beginMeasuring() {
+    HAL_GPIO_WritePin(CS_GPIO_PORT, CS_PIN, GPIO_PIN_SET);
+    reset();
+    setLowNoiseMode();
+    calibrate();
+}
+
+void ICM42688::setBank(uint8_t bank) {
+    writeRegister(REG_BANK_SEL, bank);      //bank should always be 0 for acc and gyr data
+}
+
+void ICM42688::reset() {
+    setBank(0);
+    writeRegister(UB0_REG_DEVICE_CONFIG, 0x01);
+    HAL_Delay(1);
+}
+
+void ICM42688::setLowNoiseMode() {
+    writeRegister(UB0_REG_PWR_MGMT0, 0x0F);
+}
+
+void ICM42688::setGyroFS(uint8_t fssel) {
+    uint8_t reg;
+    
+    setBank(0);
+    
+    //Read current register value
+    readRegister(0x4F, 1, &reg);
+
+    //Only change FS_SEL in reg
+    reg = (fssel << 5) | (reg & 0x1F);
+
+    writeRegister(0x4F, reg);
+    
+    gyro_scale = (2000.0f / (float)(1 << fssel)) / 32768.0f;
+    gyroFS = fssel;
+}
+
+void ICM42688::calibrate() {
+    //Set at a lower range (more resolution since IMU not moving)
+    const uint8_t current_fssel = gyroFS;
+    setGyroFS(0x03);        //Set to 250 dps
```

Comment:
magic number 0x03. Should this be a #define or constexpr?

Reply:
Ah thanks, you're right; I changed it to be a #define in the latest commit.